### PR TITLE
GH-41608: [C++][Python] Extends the add_key_value to parquet::arrow and PyArrow

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

The previous pr ( https://github.com/apache/arrow/pull/34889 ) add a `AddKeyValueMetadata` to FileWriter. And now we should export it to Parquet Arrow and Python API.

### What changes are included in this PR?

1. Add `AddKeyValueMetadata` in parquet::arrow
2. Add `add_key_value_metadata` in pyarrow
3. testing

### Are these changes tested?

Yes

### Are there any user-facing changes?

New api allowing add key-value metadata to Parquet file

* GitHub Issue: #41608

### CODE REVIEW

File: cpp/src/parquet/type_fwd.h [hunk 1]
```diff
@@ -78,9 +78,7 @@ class ReaderProperties;
 class ArrowReaderProperties;
 
 class WriterProperties;
-class WriterPropertiesBuilder;
```
Comment (mapleFU) [inline cpp/src/parquet/type_fwd.h:81]: We don't actually have this class, lol

File: cpp/src/parquet/arrow/writer.h [hunk 2]
```diff
@@ -143,6 +143,16 @@ class PARQUET_EXPORT FileWriter {
...
 
   virtual MemoryPool* memory_pool() const = 0;
+  /// \brief Add key-value metadata to the file.
+  /// \param[in] key_value_metadata the metadata to add.
+  /// \note This will overwrite any existing metadata with the same key.
+  /// \return Error if Close() has been called.
+  ///
+  /// WARNING: If `store_schema` is enabled, `ARROW:schema` would be stored
+  /// in the key-value metadata. Overwriting this key would result in
+  /// `store_schema` unusable during read.
```
Comment (wgtmac) [inline cpp/src/parquet/arrow/writer.h:153]:
```suggestion
  /// undefined behavior of `ARROW:schema` during read.
```

Comment (mapleFU) [reply cpp/src/parquet/arrow/writer.h:153]: Seems this cannot be used here? And if schema cannot being read, it would use "physical" schema?

**Comment by mapleFU** [comment]
I've no idea why R language CI failed

File: cpp/src/parquet/type_fwd.h (hunk 1 above)
Comment (mapleFU) [reply cpp/src/parquet/type_fwd.h:81]: Sigh, maybe this is for R language ( see `r/src/parquet.cpp`), revert it back

**Comment by mapleFU** [comment]
@AlenkaF @jorisvandenbossche @pitrou Would you mind take a look?

**Review by AlenkaF — APPROVED** [review]
Thank you for adding this!

Python/Cython part LGTM 👍 with only one suggestion: I think the python test fits better in tests/parquet/test_parquet_writer.py.

**Comment by mapleFU** [comment]
Migrate to `test_parquet_writer.py` now. @pitrou would you mind revisit C++ part or merge this?

File: python/pyarrow/_parquet.pxd [hunk 3]
```diff
@@ -24,7 +24,7 @@ from pyarrow.includes.libarrow cimport (CChunkedArray, CScalar, CSchema, CStatus
                                         CKeyValueMetadata, CRandomAccessFile,
                                         COutputStream, CCacheOptions,
                                         TimeUnit, CRecordBatchReader)
-from pyarrow.lib cimport _Weakrefable
+from pyarrow.lib cimport (_Weakrefable, pyarrow_unwrap_metadata, KeyValueMetadata)
```
Comment (pitrou) [inline python/pyarrow/_parquet.pxd:27]: Note that the new imports are only required in `_parquet.pyx` AFAICT.

File: python/pyarrow/parquet/core.py [hunk 4]
```diff
@@ -1108,6 +1108,18 @@ def close(self):
...
             self.file_handle.close()
 
+    def add_key_value_metadata(self, key_value_metadata):
+        """
+        Add key-value metadata to the file.
+        This will overwrite any existing metadata with the same key.
+
+        Parameters
+        ----------
+        key_value_metadata : {Key, Value}
```
Comment (pitrou) [inline python/pyarrow/parquet/core.py:1118]: Let's reuse the same conventions as in other docstrings:
```suggestion
        key_value_metadata : dict
            Keys and values must be string-like / coercible to bytes
```

**Review by pitrou — APPROVED** [review]
LGTM, just two nits

**Comment by mapleFU** [comment]
@pitrou comment fixed

**Comment by pitrou** [comment]
@mapleFU Feel free to merge if CI is fine.

**Comment by mapleFU** [comment]
CI failed is unrelated, merge

**Pull request merged**